common/src/main/java/org/apache/uniffle/common/netty/TransportFrameDecoder.java

```diff
@@ -0,0 +1,191 @@
+ * interceptor. When the interceptor indicates that it doesn't need to read any more data, framing
+ * resumes. Interceptors should not hold references to the data buffers provided to their handle()
+ * method.
+ */
+public class TransportFrameDecoder extends ChannelInboundHandlerAdapter implements FrameDecoder {
+  private int msgSize = -1;
+  private Message.Type curType = Message.Type.UNKNOWN_TYPE;
+  private ByteBuf headerBuf = Unpooled.buffer(HEADER_SIZE, HEADER_SIZE);
+  private static final int MAX_FRAME_SIZE = Integer.MAX_VALUE;
+  private static final int UNKNOWN_FRAME_SIZE = -1;
+
+  private final LinkedList<ByteBuf> buffers = new LinkedList<>();
+
+  private long totalSize = 0;
+  private long nextFrameSize = UNKNOWN_FRAME_SIZE;
+
+  @Override
+  public void channelRead(ChannelHandlerContext ctx, Object data) {
+    ByteBuf in = (ByteBuf) data;
+    buffers.add(in);
+    totalSize += in.readableBytes();
+
+    while (!buffers.isEmpty()) {
+      ByteBuf frame = decodeNext();
+      if (frame == null) {
+        break;
+      }
+      Message msg = Message.decode(curType, frame);
+      frame.release();
+      ctx.fireChannelRead(msg);
+      clear();
+    }
+  }
+
+  private void clear() {
+    curType = Message.Type.UNKNOWN_TYPE;
+    msgSize = -1;
+    headerBuf.clear();
+  }
+
+  private long decodeFrameSize() {
+    if (nextFrameSize != UNKNOWN_FRAME_SIZE || totalSize < HEADER_SIZE) {
+      return nextFrameSize;
+    }
+
+    // We know there's enough data. If the first buffer contains all the data, great. Otherwise,
+    // hold the bytes for the frame length in a composite buffer until we have enough data to read
+    // the frame size. Normally, it should be rare to need more than one buffer to read the frame
+    // size.
+    ByteBuf first = buffers.getFirst();
+    if (first.readableBytes() >= HEADER_SIZE) {
+      msgSize = first.readInt();
+      curType = Message.Type.decode(first);
+      nextFrameSize = msgSize;
+      totalSize -= HEADER_SIZE;
+      if (!first.isReadable()) {
+        buffers.removeFirst().release();
+      }
+      return nextFrameSize;
+    }
+
+    while (headerBuf.readableBytes() < HEADER_SIZE) {
+      ByteBuf next = buffers.getFirst();
+      int toRead = Math.min(next.readableBytes(), HEADER_SIZE - headerBuf.readableBytes());
+      headerBuf.writeBytes(next, toRead);
+      if (!next.isReadable()) {
+        buffers.removeFirst().release();
+      }
+    }
+
+    msgSize = headerBuf.readInt();
+    curType = Message.Type.decode(headerBuf);
+    nextFrameSize = msgSize;
+    totalSize -= HEADER_SIZE;
+    return nextFrameSize;
+  }
+
+  private ByteBuf decodeNext() {
+    long frameSize = decodeFrameSize();
+    if (frameSize == UNKNOWN_FRAME_SIZE || totalSize < frameSize) {
+      return null;
+    }
+
+    // Reset size for next frame.
+    nextFrameSize = UNKNOWN_FRAME_SIZE;
+
+    Preconditions.checkArgument(frameSize < MAX_FRAME_SIZE, "Too large frame: %s", frameSize);
```

Comment:
It seems that we can `frameSize <= MAX_FRAME_SIZE`?

Reply:
Maybe not very important, Spark does the same.